File: СупМод/Лр_1/Код/variant1.py
import numpy as np
from datetime import datetime
from time import time

def benchmark(fn, nrepeat, *args):
	start = time()
	for i in range(nrepeat):
		res = fn(*args)
	end = time()
	ms_per_run = (end - start) * 1000 / nrepeat
	return res, ms_per_run

def quadratic_form(A,x):
	n = x.shape[0]
	xTa = np.zeros(n)
	for i in range(n):
		sum = 0
		for j in range(n):
			sum += x[j] * A[j][i]
		xTa[i] = sum
	result = 0
	for i in range(n):
		result += xTa[i] * x[i]
	return result

	

n = int(input())
a = []
for i in range(n):
	a_i = list(map(float, input().split(' ')))
	if len(a_i) == n:
		a.append(a_i)
	else:
		print('Значения матрицы заполнены неверно!')
		exit()
x = list(map(float, input().split(' ')))
if len(x) != n:
	print('Значения вектора заполнены неверно!')
	exit()


np_A = np.array(a)
np_x = np.array(x)
res, time = benchmark(quadratic_form, 1000, np_A, np_x)
print(f'timing: {time}')
print(f'result: {res}')

File: СупМод/Лр_1/Код/variant1.cpp
#include <iostream>
#include <memory>
#include <cstdint>
#include <iomanip>
#include "array_types.hpp"

#include <functional>
#include <chrono>

using intptr_t = std::intptr_t;

template <class T>
struct benchresult {
    T result;
    double btime;
};

template <class T, class input_type>
auto benchmark(std::function<T(input_type)> fn, input_type input, intptr_t nrepeat){
    T result;
    auto start = std::chrono::steady_clock::now();
    for (intptr_t i = 0; i < nrepeat; i++) {
        result = fn(input);
    }
    auto end = std::chrono::steady_clock::now();
    std::chrono::duration<double> duration_s = end - start;
    double ms_per_run = duration_s.count() * 1000 / nrepeat;
    return benchresult<T> {result, ms_per_run};
}

template <class T>
T quadratic_form(matrix<T> A, vec<T> x)
{   
	intptr_t N = x.length();
	vec<T> xTa(N);
	T sum = 0;
	for (intptr_t i = 0; i < N; i++) {
		xTa(i) = 0;
		for (intptr_t j = 0; j < N; j++) {
			xTa(i) += x(j) * A(j, i);
		}
		sum += xTa(i) * x(i);
	}
	return sum;
}

int main(int argc, char* argv[])
{
    size_t n;

    std::cin >> n;
    matrix<double> a(n, n);
    vec<double> x(n);
    
    for (intptr_t k = 0; k < n * n; k++) {
        std::cin >> a(k);
    }

    for (intptr_t k = 0; k < n; k++) {
        std::cin >> x(k);
    }

    std::function<double(int)> test_quadratic = [=](int idx) {return quadratic_form(a, x);};

    auto benchresult = benchmark(test_quadratic, 0, 1000);

    std::cout << std::setprecision(std::numeric_limits<double>::digits10 + 1)
              << "Timing: " << benchresult.btime << " ms\n"
              << "Answer = " << benchresult.result
              << std::endl;
    return 0;
}
